lua/nvim-treesitter/locals.lua

```diff
@@ -2,13 +2,104 @@
 -- Locals are a generalization of definition and scopes
 -- its the way nvim-treesitter uses to "understand" the code
 
-local queries = require "nvim-treesitter.query"
 local ts_utils = require "nvim-treesitter.ts_utils"
 local ts_query = vim.treesitter.query
+local parsers = require'nvim-treesitter.parsers'
 local api = vim.api
+local caching = require'nvim-treesitter.caching'
+local queries = require'nvim-treesitter.query'
 
 local M = {}
 
+local scope_tree = caching.create_buffer_cache()
+
+local function update_scope_tree(cache, buf, parse_node, start_row, end_row)
+  local parser = parsers.get_parser(buf)
+
+  --TODO: do that in a refining manner only updating a specific region?
+  cache.scopes = {}
```

Comment:
I think that, in order to do so, we'll have to use a better data structure, indexed by lines.

Reply:
@vigoux don't think that would work; see comment above.

Reply:
or something like this https://en.wikipedia.org/wiki/Interval_tree

Reply:
Did not know about it, but that looks like something we'd use, yes